Better name propagation by macros

ducky64 opened the pull request:
Implementation for #303.
Doesn't use it anywhere (yet), so this only adds infrastructure. Hopefully also makes this less like to break things.
A (near) future PR would add this annotation to some core utils and replace TransitName.

ucbjrl commented:
retest this please

ucbjrl commented:
retest this please

ucbjrl commented:
retest this please

colinschmidt commented:
Is this now supposed to work with Chisel._ and not just chisel3._ ?

ducky64 commented:
Yes, in theory.

ducky64 commented:
Fixed the build.sbt, moved compiler plugins and the scala.lang.reflect library deps into commonSettings.

ducky64 commented:
@sdtwigg poke

colinschmidt commented:
rocket-chip uses chisel3/224ee397db324067b35874bf22c76a63e9ca531b
and just building with this branch on rocket-chip/master results in the issue seen in https://github.com/ucb-bar/rocket-chip/pull/423

So maybe rebase this onto a more recent chisel3?

ducky64 commented:
Rebased to chisel3/224ee397db324067b35874bf22c76a63e9ca531b

colinschmidt commented:
Thanks that solved this issue in rocket-chip master.

colinschmidt commented:
So I checked this out for the if/when naming and that seems to work well, but I haven't been able to get the function annotations working yet.

My chisel code looks like:

This module 
```
@chiselName
class IBox(implicit p: Parameters) extends VMUModule()(p) {
 ...
  val op = VMUDecodedOp(opq.io.deq.bits)
```
calls this function
```
  @chiselName
  def apply(op: VMUOp)(implicit p: Parameters): VMUDecodedOp = {
    val dec = Wire(new VMUDecodedOp)
...
    dec.mt := DecodedMemType(op.fn.mt)
```
which calls this function
```
object DecodedMemType {
  @chiselName
  def apply[T <: UInt](mt: T): DecodedMemType = {
    val b = (mt === MT_B)
    val h = (mt === MT_H)
    val w = (mt === MT_W)
    val d = (mt === MT_D)
    val bu = (mt === MT_BU)
    val hu = (mt === MT_HU)
    val wu = (mt === MT_WU)

    val dec = Wire(new DecodedMemType)
    dec.b := (b || bu)
    dec.h := (h || hu)
    dec.w := (w || wu)
    dec.d := d
    dec.signed := ((b || h) || (w || d))
    dec
  } 
} 
```
I was expecting all of the val b, val h, etc. to be named in the firrtl but the output firrtl looks like:
```
    node T_2329 = eq(opqInst.io.deq.bits.fn.mt, UInt<3>("h00")) @[types-vmu.scala 71:17]                 
    node T_2330 = eq(opqInst.io.deq.bits.fn.mt, UInt<3>("h01")) @[types-vmu.scala 72:17] 
    node T_2331 = eq(opqInst.io.deq.bits.fn.mt, UInt<3>("h02")) @[types-vmu.scala 73:17]                 
    node T_2332 = eq(opqInst.io.deq.bits.fn.mt, UInt<3>("h03")) @[types-vmu.scala 74:17]
    node T_2333 = eq(opqInst.io.deq.bits.fn.mt, UInt<3>("h04")) @[types-vmu.scala 75:18]
    node T_2334 = eq(opqInst.io.deq.bits.fn.mt, UInt<3>("h05")) @[types-vmu.scala 76:18]
    node T_2335 = eq(opqInst.io.deq.bits.fn.mt, UInt<3>("h06")) @[types-vmu.scala 77:18]                 
    wire T_2342 : {b : UInt<1>, h : UInt<1>, w : UInt<1>, d : UInt<1>, signed : UInt<1>} @[types-vmu.scala 79:19]
    T_2342 is invalid @[types-vmu.scala 79:19]
    node T_2348 = or(T_2329, T_2333) @[types-vmu.scala 80:17]
    T_2342.b <= T_2348 @[types-vmu.scala 80:11]
    node T_2349 = or(T_2330, T_2334) @[types-vmu.scala 81:17]
    T_2342.h <= T_2349 @[types-vmu.scala 81:11]
    node T_2350 = or(T_2331, T_2335) @[types-vmu.scala 82:17]
    T_2342.w <= T_2350 @[types-vmu.scala 82:11]
    T_2342.d <= T_2332 @[types-vmu.scala 83:11]
    node T_2351 = or(T_2329, T_2330) @[types-vmu.scala 84:23]
    node T_2352 = or(T_2331, T_2332) @[types-vmu.scala 84:35]
    node T_2353 = or(T_2351, T_2352) @[types-vmu.scala 84:29]                                            
    T_2342.signed <= T_2353 @[types-vmu.scala 84:16]                                                     
    op.mt <- T_2342 @[vmu.scala 80:12]
```

ducky64 commented:
So the name propagation breaks here:
```
  @chiselName
  def apply(op: VMUOp)(implicit p: Parameters): VMUDecodedOp = {
    val dec = Wire(new VMUDecodedOp)
    ...
    dec.mt := DecodedMemType(op.fn.mt)
```
Because `dec.mt := DecodedMemType(op.fn.mt)` is not in the form of `val name = blah`, the naming annotation doesn't recognize it and is unable to provide a prefix for the contents of DecodedMemType.

In general, handling these cases is hard since the assignment is conditional and may only happen in some circumstances. Currently, those are just ignored as far as naming is concerned.

It may be possible to handle IO assignments that are static (top level scope, only one assignment) as a special case, since assigning it to a val, then assigning the val to a IO is pretty cumbersome. Though it's also probably very hard to detect all cases of reassignment by traversing the AST, so there will be edge cases. Thoughts?

sdtwigg commented:
Why doesn't it just default to calling the signals b, h, bu, hu?

Also, the original proposal involved two macros: name and propagateName so the user could decide if you wanted the function to just name things directly versus  naming things based on names applied to the return. (Although as noted above, even for the latter, it should still default to applying the proposed suffix.)

Similarly, if the user manually calls suggestName on the result of DecodedMemType (before assignment), will it propagate the suggested name into the internal function nodes?

ducky64 commented:
For directly naming signals in functions without a prefix: it can be problematic if the function is called multiple times in the same module, then there would be a conflict. I also think if we wanted to specify that things be named without a prefix, that would have to happen in the calling scope rather than the function being called, since (in the general case) when you're writing the function you can't know ahead of time if it's going to be used in a situation that will cause a conflict.

suggestName won't work, this doesn't use the TransitName facility. I still think the postname hooks directly in Module is inappropriate coupling and very special-case functionality. The two options are for fixing this are:
- Define it as a val and use the resulting val. This requires two lines, so the resulting code is less elegant.
- The internal naming functions could be exposed and documented as part of the API. This can be an "escape hatch" that allows names and propagation in special cases to be handled. However, the syntax would still be bulky and could mean duplication of information (for example, the above problem might be handled with `dec.mt := Named(DecodedMemType(op.fn.mt), "dec.mt")`, but writing `dec.mt` twice is undesirable.
- Maybe a better smarter solution that I haven't thought of yet?

ducky64 commented:
Moved chiselName annotations to chisel3.experimental and Chisel.experimental, rebase to latest.

ducky64 commented:
Major changes:
- Added a `@blindName` for functions. This does not need anything to be returned to function, and is recursive. No annotation chain is needed to the Module level.
- `@chiselName` still behaves the same for methods (hierarchical naming if a return value is given and used by the calling function, assuming there are `@chiselName` annotations all the way up the function call stack to the Module level.
- `@chiselName` applies `@blindName` transforms for inner methods, unless they are explicitly annotated by `@chiselName`.
- Added naming collision tests. In cases of conflict, subsequent names have the _(number) attached.
- I'm going to hold off on exposing a `Named(...)` construct because of potential interplay issues with the macro-generated hierarchical naming context system.

@sdtwigg

sdtwigg commented:
A tempting thing to suggest is that, instead of blindName, call the second annotation "suggestNames" to imply it is equivalent to calling suggestName on (almost all) nested vals.

ducky64 commented:
So the goal of blindName was to make it clear that it doesn't respect the call hierarchy and it would have higher potential for undesirable behavior (like collisions). Users should think before applying this over chiselName.

I think suggestName is pretty general; if anything it would make sense to replace chiselName (which is somewhat content-free) with suggestName.

ducky64 commented:
Our resolution is to change blindName to localName, and to leave chiselName as is. Implementation behavior will not change. Changes pending...

ducky64 commented:
Rename done, review ready.

azidar commented:
@jackkoenig please review.

ducky64 commented:
Bump.

This PR is now 3 months old...

jackkoenig commented on src/main/scala/chisel3/package.scala:
```diff
@@ -281,5 +281,26 @@ package object chisel3 {    // scalastyle:ignore package.object.name
+    @compileTimeOnly("enable macro paradise to expand macro annotations")
+    class localName extends StaticAnnotation {
+      def macroTransform(annottees: Any*): Any = macro chisel3.internal.naming.NamingTransforms.localName
+    }
```
For most of the stuff in these packages, it's just a type alias, should that not be the case here? Regardless, since these are the actual annotations, I think the ScalaDoc explaining them should live here. It's not clear looking at the ScalaDoc generated by this branch where I would find a description of chiselName or localName

ducky64 commented on chiselFrontend/src/main/scala/chisel3/internal/Builder.scala:
```diff
@@ -222,3 +224,11 @@ private[chisel3] object Builder {
+  */
+object DynamicNamingStack {
+  def apply() = Builder.namingStack
+}
```
Macros expand such that this is generated into annotated functions, so it can't be private. I also know of no way to suppress warnings.

jackkoenig commented:
So, this works:

```scala
@chiselName
class MySimpleModule extends Module {
  def addOne(in: UInt): UInt = {
    val inc = Wire(in)
    inc := in +& 1.U
    inc
  }

  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })
  io.out := addOne(io.in)
}
```

And this works:

```scala
object Utils {
  @localName
  def addOne(in: UInt): UInt = {
    val inc = Wire(in)
    inc := in +& 1.U
    inc
  }
}

class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })
  io.out := Utils.addOne(io.in)
}
```

But this errors:

```scala
@chiselName
object Utils {
  def addOne(in: UInt): UInt = {
    val inc = Wire(in)
    inc := in +& 1.U
    inc
  }
}

class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })
  io.out := Utils.addOne(io.in)
}
```

error message:
```
[error] /Users/jack.koenig3/ucb-bar/chisel3-nameprop/src/test/scala/chiselTests/NamingAnnotationTest.scala:34: @chiselName annotation did not match exactly one valid tree, got:
[error] object Utils {
[error]   def addOne(in: UInt): UInt = {
[error]     val inc = Wire(in);
[error]     inc.:=(in.+&(1.U));
[error]     inc
[error]   }
[error] }
[error] @chiselName
[error]  ^
```

(and if I just annotate the function `addOne` with `@chiselName`, nothing happens)

Is annotating objects with `@chiselName` supposed to work?

ducky64 commented:
Annotating objects is not supposed to work, you should explicitly annotate each function within.

Unlike on Modules, where you can guess that inner functions are probably helper functions that are called once, I don't think the same applies on general objects. So I prefer to leave it for the user to specify each method explicitly.

That being said, the error message could be more helpful (from a "what you did wrong" rather than "what the compiler saw that's wrong" standpoint)

jackkoenig commented:
Okay fair enough, yeah I'm not sure how easy it is to add a better error message to the macro but if possible that would be helpful.

Still possibly an issue?:
If I annotate just the function with `@chiselName` it doesn't do anything:

```scala
object Utils {
  @chiselName
  def addOne(in: UInt): UInt = {
    val inc = Wire(in)
    inc := in +& 1.U
    inc
  }
}

class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })
  io.out := Utils.addOne(io.in)
}
```

Gives
```
circuit MySimpleModule :
  module MySimpleModule :
    input clock : Clock
    input reset : UInt<1>
    output io : {flip in : UInt<32>, out : UInt<32>}

    io is invalid
    wire _T_6 : UInt<32>
    _T_6 is invalid
    node _T_8 = add(io.in, UInt<1>("h01"))
    _T_6 <= _T_8
    io.out <= _T_6
```

jackkoenig commented:
By Module methods being "helper functions called once" do you mean *at least* once or *exactly* once?

ducky64 commented:
For your example: you also need to annotate your top-level module. chiselName annotations must be unbroken in the call chain to work properly. 

Helper functions called once: no convention is enforced. The idea of localName is that functions aren't called more than once in the same scope so there's less ambiguity. But if they are, it will just resolve with the default name disambiguation: `_1`, `_2`, ...

jackkoenig commented:
```scala
object Utils {
  @chiselName
  def addOne(in: UInt): UInt = {
    val inc = Wire(in)
    inc := in +& 1.U
    inc
  }
}

@chiselName
class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })
  io.out := Utils.addOne(io.in)
}

object Test extends App {
  chisel3.Driver.execute(args, () => new MySimpleModule)
}
```

So in this case, `MySimpleModule` is the top-level and it still doesn't seem to be working. Anything else I'm missing?

ducky64 commented:
So it also only recognizes direct val assign (`val myNamedThingy = something`) statements, it doesn't recognize Chisel assignment. The reasoning for this is that Chisel assignments aren't final, so in this case I can't infer that the result of Utils.addOne(...) will be io.out in all cases.

jackkoenig commented:
The idea was the `inc` would get propagated, which it does with `@localName` or if `addOne` is a method on the Module class, but does not work in the above example with `@chiselName`

ducky64 commented:
I don't think it makes sense to infer a prefix when a complete chain doesn't exist for chiselName, since the idea of chiselName is that you want that prefix. For example, if you called addOne many times in your module, then you would just get spammed with incs trailed by numbers.

sdtwigg commented:
A  series of inc_1, inc_2, inc_3 would definitely be preferable to W25, W35, W45.

ducky64 commented:
I see your point, though missing the context is also undesirable.

A possible change would be to have chiselName generate the longest contiguous prefix (from named node) of chiselName chained functions. This would also eliminate localName (which is equivalent to chiselName with no enclosing elements in the chain). The trade-off here is that it would generate more helpful, but possibly less consistent names that may change as code is moved around. One of the hopes was the chiselName was strict enough that you could use the name in downstream tools (for example, constraining module placement) and it would remain consistent through some types of refactoring (for example, moving blocks of code around, where the numerical suffix could change).

What are everyone else's thoughts?

jackkoenig commented:
I think I'm finally starting to understand chiselName. So in the following:

```scala
object Utils {
  @chiselName
  def addTwo(in: UInt): UInt = {
    val inc = in +& 1.U
    inc +& 1.U
  }
}

@chiselName
class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })
  val sum = Utils.addTwo(io.in)
  io.out := sum
}
```
I get the following Firrtl
```
circuit MySimpleModule :
  module MySimpleModule :
    input clock : Clock
    input reset : UInt<1>
    output io : {flip in : UInt<32>, out : UInt<32>}

    io is invalid
    node sum_inc = add(io.in, UInt<1>("h01")) 
    node sum = add(sum_inc, UInt<1>("h01")) 
    io.out <= sum 
```
Pretty cool. I guess the thing that concerns me is that if I instead connect to the result of `addTwo` directly, ie.

```scala
  io.out = Utils.addTwo(io.in)
```

I all of the sudden lose everything:

```
    node _T_7 = add(io.in, UInt<1>("h01"))
    node _T_9 = add(_T_7, UInt<1>("h01"))
    io.out <= _T_9
```

I see what you mean about trying to provide guarantees, but it seems strange to me that everything should just get thrown away if it's not a perfect line all the way to the top. As a library writer, I would like for people to get decent names even if they're unwilling or unable to write the right code to create these chains. I think your idea of providing the longest unbroken chain is a good one, since you can have fairly sophisticated libraries generating large chains with nice names and yet have a user invoke the library in a bad way. At a bare minimum, this would make the naming so much more stable that I think the benefit there is worth the possibility that people may use libraries wrong and not get perfect naming guarantees.

ducky64 commented:
How about this proposal:
- Make chiselName generate names according to longest unbroken chain
- Eliminate localName (it's redundant)
- No guarantees at this point in time, it can be looked into when there's a use case
  - (+) Practically users have little control over library writers and whether they provide a unbroken chiselName. The guarantee semantics may needs to be encoded elsewhere (not in naming annotation in libraries)
  - (-) If guarantees require changing the naming annotation that becomes more difficult if everyone uses chiselName.

jackkoenig commented:
I think this proposal makes a lot of sense

colinschmidt commented:
I think the name guarantee idea can be handled two ways. Either you can provide some additional annotation that hardcodes a name and errors during elaboration if it can't get that name. Or you could provide some sort of firrtl annotation that the user can consume in their backend.

Otherwise I like the elimination of the second macro and the streamlining this proposal seems to add.

ducky64 commented:
Updated. localName eliminated. chiselName does best-effort (and will drop non-present links in the chain, instead of maximal contiguous chain, since some prefix is probably better than no prefix).

Note that the chiselName annotation still must be applied at the top level. Eventually, it might be rolled into one of the things that happens on a Module close, but chiselName is purely opt-in for now (in case something goes wrong).

jackkoenig commented:
This looks good. With the recent changes, would it make sense to allow chiselName on objects as well so that you could annotate just a utilities object and not have to do it on every single function? It could also annotate any modules in side for you as well. Would this work?

jackkoenig commented:
Also, this works on `if` blocks (thanks!!!!) but doesn't seem to work on for loops:

```scala
// This does NOT work!
@chiselName
class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })

  for (i <- 0 until 1) {
    val a = Wire(init = io.in)
    val b = Wire(init = a)
    io.out := b
  }
}

// This does work!
@chiselName
class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })

  if (true) {
    val a = Wire(init = io.in)
    val b = Wire(init = a)
    io.out := b
  } else {
    io.out := func(io.in)
  }
}

```

jackkoenig commented:
Does not work on anonymous functions either, for example:

```scala
@chiselName
class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })

  Seq(io.in).foreach { in =>
    val a = Wire(init = in)
    val b = Wire(init = a)
    io.out := b
  }
}
```

If supporting all of these cases is extremely difficult then obviously we can move ahead without them, just wanted to check.

sdtwigg commented:
scala desugars for loops into a bunch of function calls with the inner body inside the parameters so I wonder if `case q"(..$params) => $expr" => tree` is blocking that. (Interestingly, even though it does that desugaring really early in the compile process, quasiquotes still has matchers for 'for' and 'for-yield'. I wonder if it only intends for you to use those to build, not match, trees.)

jackkoenig commented:
Just for fun, it **does** work with match:

```scala
@chiselName
class MySimpleModule extends Module {
  val io = IO(new Bundle {
    val in = Input(UInt(32.W))
    val out = Output(UInt(32.W))
  })

  io.in match {
    case in =>
      val a = Wire(init = in)
      val b = Wire(init = a)
      io.out := b
  }
}
```

ducky64 commented:
Yeah, so the for loop is turning into (using the new `@dump` debugging annotation)
```scala
   0.until(1).foreach(((i) => {
     val a = Wire(init = io.in);
     val b = Wire(init = a);
     io.out.$colon$eq(b)
   }));
```
which for some reason isn't matching. I'll look into it. 

As far as recursively naming inner functions / objects / whatever: it's possible (and not even difficult to implement), but I'm not sure whether it's a good idea or not since it might be desirable to limit the scope of the annotation. Theoretically, it should be harmless (performance aside, since it is executing more code) when applied on non-Chisel code. Thoughts?

ducky64 commented:
For some reason I wrote a rule preventing it from recursing into anonymous functions. That's gone now, so that and the for case should work.
It will be likely to produce name collisions though, but a future change may be to introduce some kind of variable interpolation syntax into the variable name, like `a_i` which would turn into `a_0`, `a_1`

jackkoenig commented:
I guess it would be nice to have a test for for loops and anonymous functions, but yeah this looks great. I'm fine with the decision to not do the recursive macro applying, seems reasonable enough. People might request it in the future but we can cross that bridge if/when we come to it!

ducky64 commented:
Tests added! for loop test is pretty basic because it's difficult to get the expected disambiguated name in without significant plumbing, but the anonymous function test is more robust.

jackkoenig commented:
Looks great!
